TinyPG runnable as command line

Jissai opened the pull request:
*add command line support on Main()
*Refactor and isolate business code (Parse/build/eval) from View (MainForm)
*add unit test

Jissai committed:
TinyPG callable from command line

* refactor and isolate business code (aka parse/buil/eval) from
presentation (MainForm)
* add commandline support on Main() with branch to either GUI or Silent
run

Jissai committed:
TinyPG callable from command line

strip the StringBuilder from method parameters, it now stand as a
Program Property

Jissai committed:
TinyPG callable from command line

*bug correction
*add unit test
*code refactoring

Jissai committed:
Merge branch 'refs/heads/bugfix2'

Jissai committed:
TinyPG callable from command line

*add some unit test

tomspilman commented on TinyPG.UnitTests/ProgramTester.cs:
```diff
@@ -0,0 +1,79 @@
+    /// Summary description for ProgramTester
+    /// </summary>
+    [TestClass]
+    public class ProgramTester
```
What is this class for?

The `<summary>` is basically empty.

Is it just a unit test for your Program class changes?  Or is it more than that?

Jissai replied:
Hi Tom,

Didn't have access to the internet for the week end. I will try to answer
to all your questions.

1/ in MainForm.cs there is a lot of deleted code. I placed the code
in Programs.cs to reduce code duplication : Parse/build/eval of grammar is
done only in Program now and Mainform.cs beiing the view, it should not
contains any business methods.

2/ Sorry for the spaces, I done a code auto-formatting at one time and even
if i tried to go back with the same number of spaces, git report them :(

3/ //var args = Environment.GetCommandLineArgs();
Sure you can remove it, I wasn't sure to reproduce the same behavior of
this command before i changed Main() 's signature, so I commented it for
reference "just in case". I just forget to remove it.

4/ I thought it was for the best... well apparently not. will do as you say
next time :)

5/ Yes ProgramTester is there to help test Program.cs. I should have filled
the <summary>



I think I addressed all your mails. Thank you for the reviews and feedbacks.
If any questions, fell free to ask.

Jissai

2013/3/8 Tom Spilman notifications@github.com

> In TinyPG.UnitTests/ProgramTester.cs:
> 
> > @@ -0,0 +1,79 @@
> > +﻿using System;
> > +using System.Text;
> > +using System.Collections.Generic;
> > +using System.Linq;
> > +using Microsoft.VisualStudio.TestTools.UnitTesting;
> > +using TinyPG.Compiler;
> > +
> > +namespace TinyPG.UnitTests
> > +{
> > -    /// <summary>
> >   
> >   
> > -    /// Summary description for ProgramTester
> > -    /// </summary>
> >   
> >   
> > -    [TestClass]
> > -    public class ProgramTester
> 
> What is this class for?
> 
> The <summary>
> 
>  is basically empty.
> 
> Is it just a unit test for your Program class changes? Or is it more than
> that?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SickheadGames/TinyPG/pull/9/files#r3301515
> .

Jissai replied:
done in next commit

tomspilman commented on TinyPG/Compiler/Grammar.cs:
```diff
@@ -155,8 +155,6 @@ private void SetupDirectives()
                 d["OutputPath"] = "./"; // write files to current path
             if (!d.ContainsKey("Language"))
-                d["OutputPath"] = "C#"; // write files to current path
-            if (!d.ContainsKey("Language"))
```
Notice how this is in here and in your other pull request #8?

This happens all the time with people new to git.  You should never change the master branch directly...  instead _always_ make a feature branch for implementing new features.  You then submit that branch in the PR and it will eventually make it back to your master branch.

If you need you features/fixes for yourself before they are accepted in a PR the best thing to do is have your own "stable" branch.  You can merge your feature and fix branches into that directly.

tomspilman commented on TinyPG/Program.cs:
```diff
@@ -8,27 +8,133 @@
-            Application.SetCompatibleTextRenderingDefault(false);
-            Application.Run(new MainForm());
-            return;
+            //var args = Environment.GetCommandLineArgs();
```
Can you remove this commented out line?

Jissai replied:
done in next commit

tomspilman commented on TinyPG/MainForm.cs:
```diff
@@ -152,13 +151,13 @@ private void MainForm_Load(object sender, EventArgs e)
                 || e.NewContext.Token.Type == TinyPG.Highlighter.TokenType.GRAMMARSTRING
                 || e.NewContext.Token.Type == TinyPG.Highlighter.TokenType.DIRECTIVESTRING
-                || e.NewContext.Token.Type == TinyPG.Highlighter.TokenType.GRAMMARCOMMENTBLOCK 
+                || e.NewContext.Token.Type == TinyPG.Highlighter.TokenType.GRAMMARCOMMENTBLOCK
```
For future reference you can should be careful to avoid these unnecessary whitespace changes.  

Git can still generate conflicts when merging on trivial stuff like whitespace.

tomspilman commented on TinyPG/MainForm.cs:
```diff
@@ -572,9 +571,17 @@ private void SetHighlighterLanguage(string language)
+            textEditor.Select(tree.Errors[0].Position, tree.Errors[0].Length > 0 ? tree.Errors[0].Length : 1);
+        }
+
         private void CompileGrammar()
```
There is a lot of deleted code in this method.  Can you explain to me a little about what is going on here, so I can review it better?

Jissai replied:
I moved a part of the code in Programs.cs to reduce code duplication : Parse/build/eval is done only in Program.cs now and Mainform.cs should not contains any business methods, it should be only View.

It's not as clean at it should be, but the changes are already heavy.

tomspilman commented:
This is fantastic!  This mostly covers my needs for issue #3.

Just going to review your code some first then we can get this merged.

Jissai committed:
cleaning for the command line merge

tomspilman commented:
I didn't forget about you here...  it just has been a busy week.  I will get this merged in the next 24 hours.

tomspilman commented:
Looks like we may have a bug with these changes.

If I open one of my existing TPG files with the TinyPG editor and hit Generate the ParseTree.cs, Parser.cs, and Scanner.cs files are no longer generated.

Can you give that a test on your end?

Jissai commented:
I use the editor and cmd line every day with different tpg file without
problems.
I had a similar bug (and a fix) when the grammar contains errors. Is this
your case ?

Jissai committed:
BugFix :  grammar file not generated on build error

Scanner, Parser, ParserTree were not Generated if the grammar file
contains build errors

tomspilman commented:
> I had a similar bug (and a fix) when the grammar
> contains errors. Is this your case ?

Well....  yes and no.

The grammar has no errors and a correct parse tree is generated:

```
Parse successful!
```

There are however errors when the parser code is compiled....

```
Building code...
The type or namespace name 'ShaderInfo' could not be found (are you missing a using directive or an assembly reference?) on line 220
The type or namespace name 'TechniqueInfo' could not be found (are you missing a using directive or an assembly reference?) on line 230
The type or namespace name 'ShaderInfo' could not be found (are you missing a using directive or an assembly reference?) on line 241
The type or namespace name 'PassInfo' could not be found (are you missing a using directive or an assembly reference?) on line 253
The type or namespace name 'PassInfo' could not be found (are you missing a using directive or an assembly reference?) on line 261
The type or namespace name 'TechniqueInfo' could not be found (are you missing a using directive or an assembly reference?) on line 270
The type or namespace name 'PassInfo' could not be found (are you missing a using directive or an assembly reference?) on line 279
The type or namespace name 'PassInfo' could not be found (are you missing a using directive or an assembly reference?) on line 287
The type or namespace name 'SamplerStateInfo' could not be found (are you missing a using directive or an assembly reference?) on line 298
The type or namespace name 'SamplerStateInfo' could not be found (are you missing a using directive or an assembly reference?) on line 314
The type or namespace name 'ShaderInfo' could not be found (are you missing a using directive or an assembly reference?) on line 320
Compilation contains errors, could not compile.
```

These are all expected as my parser references classes and method that exist outside of the TPG file.

I guess the key thing is that this all worked fine before...  it does not now....  something changed and alter the behavior.

tomspilman commented:
So testing with your latest change seemed to fix that.

Next issue I ran into if I do this:

```
TinyPG.exe MGFX.tpg
```

The command line crashes.

If I do this....

```
TinyPG.exe .\MGFX.tpg
```

It does not.

Seems like it assumes the path contains a directory and doesn't account for the case where it doesn't.

Jissai committed:
BugFix : relative filename causes crash in cmd window

When given e relative filename, TinyPG crashes  when launched from
command line in cmd window.
SetCurrentDirectory should not be called with null or string.Empty
directory

Jissai committed:
BugFix : relative filename causes crash in cmd window

use GetFullPath on the argument to convert relative path to absolute
pat. It prevents a lot of troubles afterward.

tomspilman commented:
Ok... that seems to have fixed it.

As you have noted the console output isn't working from the command line.

I'm fine merging this now if you are and get further fixes later.

Jissai commented:
Sure we can merge it.
Thank you for your feedbacks and debugguing by the way.
Concerning the console output, the workaround i described in issue #18 works fine in my environement for now.
